registry/storage/metadata/metadata.go

```diff
@@ -0,0 +1,59 @@
+package metadata
+
+import (
+	"fmt"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/reference"
+)
+
+type ErrKeyNotFound struct {
+	Key string
+}
+
+func (err ErrKeyNotFound) Error() string {
+	return fmt.Sprintf("metadata key not found: %s", err.Key)
+}
+
+// Metadatable signifies metadata support
+type Metadatable interface {
+	MetadataService() MetadataService
+}
+
+type MetadataUpdateRecord struct {
+	Actual   interface{}
+	Expected interface{}
```

Comment:
Should `Expected` be optional? I suspect that in many cases, for example when you call `Transaction.Update`, you just want to set a value, and you don't care what value was there before. `Transaction.Update` has to do a `Get` before it can add an entry to the transaction record. This means `Transaction.Update` costs a network round trip, even though you haven't written anything yet.